### DISCUSSION TITLE
Guide to change database configuration.

### DISCUSSION THREAD

**Discussion by stockarea** [fleetbase/fleetbase#143] [Q&A] [answered]
I have my own database hosted at RDS and want to use that rather than MySQL docker container. Can someone help and guide me how to configure that in docker rather than using the docker instance. I am new to docker and little confused as where to do it.

**Comment by shivthakker** [comment]
Hi @stockarea , we do have support packages for implementation. You can schedule some time with me here:  https://calendly.com/shiv-10

**Comment by roncodes** [comment, marked as answer]
> I have my own database hosted at RDS and want to use that rather than MySQL docker container. Can someone help and guide me how to configure that in docker rather than using the docker instance. I am new to docker and little confused as where to do it.

We recommend configuring your external database, such as Amazon RDS, by setting the `DATABASE_URL` environment variable to a valid MySQL connection string. Alternatively, you can specify the database connection details by individually setting the following environment variables: `DB_HOST`, `DB_DATABASE`, `DB_USERNAME`, `DB_PASSWORD`, and `DB_PORT`.

**Reply by stockarea** [reply]
Where can I specify the environment variables? In api/.env file?
Also since my localhost is already have MySQL installed, I don't want database service to be started from docker and rather just listen to remote database server. So for that how to tell docker to not start database service and just listen to remote database connection?

**Reply by roncodes** [reply]
You have two main options for specifying environment variables when running your containers on AWS:

1. **AWS ECS Task Definitions:** If you're using Amazon ECS to manage your containers, you can define environment variables within your task definition.
2. **Parameter Store:** For more sensitive data, you might consider using AWS Systems Manager Parameter Store and referencing these parameters in your task definition or directly within your application.

For local development or testing, you can specify them in your `api/.env` file or directly in your `docker-compose.yml` or `docker-compose.override.yml` files under the environment section.

As for disabling the database service you also have two options:

If you already have MySQL running elsewhere or via RDS  and want to prevent Docker from spinning up an additional database container, there are two ways you can achieve this:

1. **Comment Out the Database Service:** You can simply comment out the database service definition in your AWS ECS task definition or in your docker-compose.yml file, if you are using one for local development.
2. **Use docker-compose.override.yml for Local Development:** For local development, if you don't want to modify the original docker-compose.yml file, you can create a docker-compose.override.yml file with the following content:
```yml
version: '3.1'
services:
  database:
    entrypoint: /bin/echo "Database service disabled"
```
By doing this, the database service will be overridden to essentially do nothing when running Docker Compose, effectively disabling it. Meanwhile, you can set your application to connect to a MySQL instance running on RDS or another remote server by using the appropriate environment variables (`DATABASE_URL` or `DB_HOST`, `DB_DATABASE`, `DB_USERNAME`, `DB_PASSWORD`, and `DB_PORT`).

**Reply by stockarea** [reply]
Thank you so much. I commented out database service and put up environment variables in api/.env and docker-compose.yml files and it is getting started. Thanks for this help


One last help needed. how can I connect to the database which is running on my local but outside docker. I know the reason of 127.0.0.1 that docker is using its own localhost server which dont' have my MySQL up and running so how can I solve this? how can I connect to my own localhost database.

**Reply by roncodes** [reply]
You're correct in noting that using 127.0.0.1 or localhost inside a Docker container will point to the container's own local network, not the host machine where your MySQL is running. Here's how you can connect to your local MySQL instance from a Docker container:

1. **Use Host Network Mode (Linux Only):** If you are running Docker on a Linux machine, you can start your container with host network mode by adding `network_mode: "host"` to your service in the `docker-compose.yml` file. This will allow the container to access the host's network stack.
```yml
services:
  your_api_service:
    network_mode: "host"
```
2. **Use a Special Hostname:** Regardless of your OS, Docker provides a special DNS name host.docker.internal that resolves to the internal IP address used by the host. In your connection string or DB_HOST environment variable, replace localhost or 127.0.0.1 with host.docker.internal.
```yml
services:
  your_api_service:
    environment:
      - DB_HOST=host.docker.internal
```
Or in your api/.env file:
```yml
DB_HOST=host.docker.internal
```
3. **Use the Host IP Address:** Same steps as 2, but you use the actual IP address of the host machine or in your case the EC2 container.

**Reply by stockarea** [reply]
Thanks you so much for this

**Comment by dsonifullstackdev** [comment]
It allow connection with no password with public IP of server where I setup mysql but when I setup password, its giving error.

**Reply by dsonifullstackdev** [reply]
Error is: SQLSTATE[HY000] [1045] Access denied for user 'mytestuser'@'mypublicip' (using password: YES)

**Reply by roncodes** [reply]
This isn't enough detail to help, you need to at minimum provide your environment/ server host if any, your setup method and the specific error.

**Reply by roncodes** [reply]
You should attempt to connect using `mycli` or `mysql-cli` to confirm your credentials are working and connectivity to the database is possible

**Reply by dsonifullstackdev** [reply]
I am setting up on digital ocean droplet using docker but with docker db is crashes in 4 to 5 hours, so I Install mysql manually and setup it listen on public IP.

**Reply by dsonifullstackdev** [reply]
Yes, tested my cretetials are working fine, i tested using command 
mysql -u mytestuser -p -h myserverip -D mydbname
DB also connected with workbench on my local system using same password

**Reply by roncodes** [reply]
> I am setting up on digital ocean droplet using docker but with docker db is crashes in 4 to 5 hours, so I Install mysql manually and setup it listen on public IP. 

For a digital ocean droplet I highly recommend using a managed database. We discovered depending on the image you use with digital ocean to the database container will crash which we don't know why yet.

Opting to setup a database on the droplet host manually is also a good option but you have to make sure the container is able to connect to the database. There should be some docker docs on this. 

Still I recommend using their managed database for an easier setup.

**Reply by roncodes** [reply]
> Yes, tested my cretetials are working fine, i tested using command 
> 
> mysql -u mytestuser -p -h myserverip -D mydbname
> 
> DB also connected with workbench on my local system using same password
> 
> 

Were you able to connect to the database from the docker container using exec? You need to ensure the database is accessible from within the container.

**Reply by dsonifullstackdev** [reply]
> > Yes, tested my cretetials are working fine, i tested using command
> > mysql -u mytestuser -p -h myserverip -D mydbname
> > DB also connected with workbench on my local system using same password
> 
> Were you able to connect to the database from the docker container using exec? You need to ensure the database is accessible from within the container.

No, its not connecting within container, giving erroras above mentioned, however its working fine when I setup no password

**Reply by dsonifullstackdev** [reply]
I setup two user with same privileges' on mysql, one with password and another with no password. 
Fleatbase perfectly connected with no password but giving error with password user

**Reply by roncodes** [reply]
This error just means the password is wrong.

**Reply by dsonifullstackdev** [reply]
> This error just means the password is wrong.

I tried by changing the password but giving same error

**Reply by dsonifullstackdev** [reply]
Looks like there is restrictions on fleetbase, it allow only with no password, anyone here used fleetbase using Docker but external database? If yes, than I will struggle more.

**Reply by roncodes** [reply]
There is several setups now on AWS, GCS, DO and other cloud providers. We don't implement any restrictions on database connections, it's just laravel framework - nothing special here. 

In fact we personally use AWS/Fargate w/ RDS setup. 

Please consult further documentation to resolve your issue.

**Reply by dsonifullstackdev** [reply]
Thanks for confirmation, I will do more try.
